From a6d30fdbfaa9c30beca5c566636d5a3fc4e945e4 Mon Sep 17 00:00:00 2001
From: Victor Seva <linuxmaniac@torreviejawireless.org>
Date: Mon, 3 Jun 2013 11:22:04 +0200
Subject: [PATCH 1/2] core: Add pv_cache_dump function.

Returns a list of pv_elem_t with the content of pv_cache.
---
 pvapi.c |   35 +++++++++++++++++++++++++++++++++++
 pvar.h  |    1 +
 2 files changed, 36 insertions(+)

--- a/pvapi.c
+++ b/pvapi.c
@@ -329,6 +329,41 @@
 	return NULL;
 }
 
+pv_elem_t* pv_cache_dump(struct sip_msg* msg) {
+	int i;
+	pv_cache_t *pvi = NULL;
+	pv_elem_t *list = NULL;
+	pv_elem_t *last = NULL;
+
+	if(_pv_cache_set==0)
+		return NULL;
+
+	for(i=0;i<PV_CACHE_SIZE;i++)
+	{
+		pvi = _pv_cache[i];
+		while(pvi)
+		{
+			if(last==NULL){
+				list = (pv_elem_t*) pkg_malloc(sizeof(pv_elem_t));
+				last = list;
+			}
+			else
+			{
+				last->next = (pv_elem_t*) pkg_malloc(sizeof(pv_elem_t));
+				last = last->next;
+			}
+			/*LM_DBG("[%d] pvar[%.*s] spec[%p]\n", i,
+				pvi->pvname.len, pvi->pvname.s,	&pvi->spec);*/
+			memset(last, 0, sizeof(pv_elem_t));
+			last->spec = &pvi->spec;
+			last->text.s = pvi->pvname.s;
+			last->text.len = pvi->pvname.len;
+			pvi = pvi->next;
+		}
+	}
+	return list;
+}
+
 /**
  *
  */
--- a/pvar.h
+++ b/pvar.h
@@ -207,6 +207,7 @@
 
 int pv_locate_name(str *in);
 pv_spec_t* pv_cache_get(str *name);
+pv_elem_t* pv_cache_dump(struct sip_msg* msg);
 
 /*! \brief PV helper functions */
 int pv_get_null(struct sip_msg *msg, pv_param_t *param, pv_value_t *res);
